File: 0134-gas-station/0134-gas-station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        
        if len(gas) == 1:
            return 0 if gas[0] >= cost[0] else -1
        
        if sum(gas) < sum(cost):
            return -1
        
        start = 0
        new_start = None
        visited_gas = start

        while start >= visited_gas:
            tank = 0
            for j in range(start, len(gas)):
                tank += gas[j]
                tank -= cost[j]
                if tank < 0:
                    new_start = j + 1 if j < len(gas) - 1 else 0
                    break
            if tank < 0:
                start = new_start
                visited_gas = max(visited_gas, start)
                continue
            for j in range(0, start):
                tank += gas[j]
                tank -= cost[j]
                if tank < 0:
                    new_start = j + 1 if j < len(gas) - 1 else 0
                    break
            if tank < 0:
                start = new_start
                visited_gas = max(visited_gas, start)
                continue
            else:
                return start
                
        return -1

File: 0134-gas-station/0134-gas-station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int totalGas = 0;
        int totalCost = 0;
        int curGas = 0;
        int start = 0;
        int n = gas.size();

        for (int i = 0; i < n; ++i) {
            totalGas += gas[i];
            curGas += gas[i];

            totalCost += cost[i];

            if (curGas < cost[i]) {
                start = i + 1;
            }
            curGas = max(0, curGas - cost[i]);
        }

        return (totalCost > totalGas) ? -1 : start;
    }
};
